{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Working with `metabolomics.info` data\n",
    "Just getting started with Jupyter Lab, like Jupyter Notebooks, but with more features.  FO rmore about Jupyter Lab see [Introduction to Jupyter Lab](https://jupyterlab.readthedocs.io/en/stable/getting_started/overview.html)\n",
    "\n",
    "In this notebook we are trying various things with python tooling for VIVO data.  Some of the cells are not VIVO specific, some are.\n",
    "\n",
    "Many of the cells use `rdflib`, a very handy library for manipulating RDF graphs using Python.  See [rdflib docs](https://rdflib.readthedocs.io/en/stable/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Query the data with Fuseki\n",
    "\n",
    "Fuseki is a stand-alone SPARQL server from the [Apache Jena Project](http://apache.org/jena).  \n",
    "Triples can be loaded into a Fuseki server and then queried.  In the code below, the triples have been loaded into\n",
    "an in-memory (very fast) database called people-portal.\n",
    "\n",
    "### Count triples in the datbase"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3.12 ms, sys: 1.6 ms, total: 4.73 ms\n",
      "Wall time: 675 ms\n",
      "1232155\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "q = \"\"\"\n",
    "SELECT (COUNT(*) AS ?ntriples)\n",
    "WHERE {\n",
    "  ?s ?p ?o .\n",
    "}\n",
    "\"\"\"\n",
    "%time response = requests.post('http://localhost:3030/people-portal/sparql', data={'query': q})\n",
    "print(response.json()['results']['bindings'][0]['ntriples']['value'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Count people, papers, organizations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2.9 ms, sys: 1.52 ms, total: 4.42 ms\n",
      "Wall time: 35 ms\n",
      "People 1372 Articles 17790 Orgs 1446\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "q = \"\"\"\n",
    "PREFIX bibo: <http://purl.org/ontology/bibo/>\n",
    "PREFIX foaf: <http://xmlns.com/foaf/0.1/>\n",
    "SELECT ?npeople ?narticles ?norgs\n",
    "WHERE {\n",
    "    \n",
    "    {\n",
    "    SELECT (COUNT(DISTINCT ?person) AS ?npeople)\n",
    "    WHERE { \n",
    "        ?person a foaf:Person .\n",
    "        }\n",
    "    }\n",
    "\n",
    "    {\n",
    "    SELECT (COUNT(DISTINCT ?article) AS ?narticles)\n",
    "    WHERE { \n",
    "        ?article a bibo:Article .\n",
    "        }\n",
    "    }\n",
    "\n",
    "    {\n",
    "    SELECT (COUNT(DISTINCT ?org) AS ?norgs)\n",
    "    WHERE { \n",
    "        ?org a foaf:Organization .\n",
    "        }\n",
    "    }\n",
    "    \n",
    "}\n",
    "\"\"\"\n",
    "%time response = requests.post('http://localhost:3030/people-portal/sparql', data={'query': q})\n",
    "vals = response.json()['results']['bindings'][0]\n",
    "print(\"People\", vals['npeople']['value'], \"Articles\", vals['narticles']['value'], \"Orgs\", vals['norgs']['value'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### People by count of papers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2.93 ms, sys: 1.57 ms, total: 4.5 ms\n",
      "Wall time: 72 ms\n",
      "Ronald Petersen 588\n",
      "Dean Jones 320\n",
      "Stefan Kaufmann 270\n",
      "Theodore Lawrence 263\n",
      "Quiying Chen 258\n",
      "Stefan H.E. Kaufmann 258\n",
      "Eva Feldman 253\n",
      "Brett Finlay 248\n",
      "Michelle Mielke 214\n",
      "John Meeker 198\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "q = \"\"\"\n",
    "PREFIX foaf: <http://xmlns.com/foaf/0.1/>\n",
    "PREFIX vivo: <http://vivoweb.org/ontology/core#>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX bibo: <http://purl.org/ontology/bibo/>\n",
    "SELECT (MAX(?lname) as ?name) (COUNT (DISTINCT ?a) AS ?npubs)\n",
    "WHERE {\n",
    "  ?person a foaf:Person .\n",
    "  ?person rdfs:label ?lname .\n",
    "  ?person vivo:relatedBy ?a .\n",
    "  ?a a vivo:Authorship .\n",
    "}\n",
    "GROUP BY ?person\n",
    "ORDER BY DESC(?npubs)\n",
    "\"\"\"\n",
    "%time response = requests.post('http://localhost:3030/people-portal/sparql', data={'query': q})\n",
    "for row in response.json()['results']['bindings'][0:10]:\n",
    "    print(row['name']['value'], row['npubs']['value'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Find co-authors\n",
    "\n",
    "For social network analysis, we are interested in finding people who are co-authors, and how many times they have been co-authors.\n",
    "For each person we look at their papers, and the authors of those papers and create a pair for the person and each of their co-authors.\n",
    "We then count the pairs.  If a is a co-author of b, then we will also find later that b is a co-author of a.  The counts will be the same."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3.95 ms, sys: 2.17 ms, total: 6.12 ms\n",
      "Wall time: 248 ms\n",
      "Stefan H.E. Kaufmann Stefan Kaufmann 257\n",
      "Stefan Kaufmann Stefan H.E. Kaufmann 257\n",
      "Michelle Mielke Ronald Petersen 138\n",
      "Ronald Petersen Michelle Mielke 138\n",
      "Dean Jones Karan Uppal 60\n",
      "Karan Uppal Dean Jones 60\n",
      "Pramod P Wangikar Pramod Wangikar 47\n",
      "Pramod Wangikar Pramod P Wangikar 47\n",
      "Dean Jones Douglas Walker 46\n",
      "Douglas Walker Dean Jones 46\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "q = \"\"\"\n",
    "PREFIX vivo: <http://vivoweb.org/ontology/core#>\n",
    "PREFIX bibo: <http://purl.org/ontology/bibo/>\n",
    "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX vitro: <http://vitro.mannlib.cornell.edu/ns/vitro/0.7#>\n",
    "PREFIX vitrop: <http://vitro.mannlib.cornell.edu/ns/vitro/public#>\n",
    "PREFIX m3c: <http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#>\n",
    "PREFIX obo: <http://purl.obolibrary.org/obo/>\n",
    "PREFIX vcard: <http://www.w3.org/2006/vcard/ns#>\n",
    "PREFIX fao: <http://aims.fao.org/aos/geopolitical.owl#>\n",
    "PREFIX foaf: <http://xmlns.com/foaf/0.1/>\n",
    "\n",
    "SELECT (MAX(?name1) AS ?name) ?p (MAX(?name2) AS ?coauthor) ?auth (COUNT(DISTINCT ?paper) AS ?npapers)\n",
    "WHERE {\n",
    "  ?p a foaf:Person .\n",
    "  ?p rdfs:label ?name1 .\n",
    "  ?p vivo:relatedBy ?a .\n",
    "  ?a a vivo:Authorship .\n",
    "  ?a vivo:relates ?paper .\n",
    "  ?paper bibo:doi ?doi .\n",
    "  ?paper vivo:relatedBy ?a2 .\n",
    "  FILTER(?a2 != ?a)\n",
    "  ?a2 vivo:relates ?auth .\n",
    "  ?auth a foaf:Person .\n",
    "  ?auth rdfs:label ?name2 .\n",
    "}\n",
    "GROUP BY ?p ?auth\n",
    "ORDER BY DESC(?npapers)\n",
    "\"\"\"\n",
    "%time response = requests.post('http://localhost:3030/people-portal/sparql', data={'query': q})\n",
    "for row in response.json()['results']['bindings'][0:10]:\n",
    "    print(row['name']['value'], row['coauthor']['value'], row['npapers']['value'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read the data into a graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1min 5s, sys: 688 ms, total: 1min 5s\n",
      "Wall time: 1min 6s\n",
      "1232155\n"
     ]
    }
   ],
   "source": [
    "from rdflib import Graph, Namespace\n",
    "from rdflib.namespace import RDF, FOAF, RDFS\n",
    "from lookup import term_lookup, VIVO, OBO, BIBO, UFL, VCARD, M3C\n",
    "\n",
    "g = Graph()\n",
    "%time g = g.parse(\"../m3c-r/m3c-20200316.nt\", format=\"nt\")\n",
    "print(len(g))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Type list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "130292 \t http://www.w3.org/2002/07/owl#Thing\n",
      "84744 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#Domain_entity\n",
      "82658 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#Dataset\n",
      "41006 \t http://purl.obolibrary.org/obo/BFO_0000002 (continuant)\n",
      "41006 \t http://purl.obolibrary.org/obo/BFO_0000001 (entity)\n",
      "20075 \t http://vivoweb.org/ontology/core#Relationship\n",
      "20075 \t http://vivoweb.org/ontology/core#Authorship\n",
      "20075 \t http://purl.obolibrary.org/obo/BFO_0000020 (specifically dependent continuant)\n",
      "17790 \t http://purl.org/ontology/bibo/Document\n",
      "17790 \t http://purl.org/ontology/bibo/Article\n",
      "17790 \t http://purl.obolibrary.org/obo/IAO_0000030 (information content entity)\n",
      "17790 \t http://purl.obolibrary.org/obo/BFO_0000031 (generically dependent continuant)\n",
      "3141 \t http://purl.obolibrary.org/obo/BFO_0000004 (independent continuant)\n",
      "2818 \t http://xmlns.com/foaf/0.1/Agent\n",
      "2529 \t http://www.w3.org/2006/vcard/ns#Identification\n",
      "2529 \t http://www.w3.org/2006/vcard/ns#Explanatory\n",
      "2529 \t http://www.w3.org/2006/vcard/ns#Communication\n",
      "2529 \t http://www.w3.org/2006/vcard/ns#Addressing\n",
      "1446 \t http://xmlns.com/foaf/0.1/Organization\n",
      "1372 \t http://xmlns.com/foaf/0.1/Person\n",
      "1372 \t http://www.w3.org/2006/vcard/ns#Name\n",
      "1070 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#Study\n",
      "910 \t http://vitro.mannlib.cornell.edu/ns/vitro/public#FileByteStream\n",
      "910 \t http://vitro.mannlib.cornell.edu/ns/vitro/public#File\n",
      "816 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#Project\n",
      "605 \t http://www.w3.org/2006/vcard/ns#Work\n",
      "605 \t http://www.w3.org/2006/vcard/ns#Type\n",
      "605 \t http://www.w3.org/2006/vcard/ns#Email\n",
      "605 \t http://www.w3.org/2006/vcard/ns#Code\n",
      "579 \t http://vivoweb.org/ontology/core#Department\n",
      "552 \t http://www.w3.org/2006/vcard/ns#Telephone\n",
      "550 \t http://vivoweb.org/ontology/core#Laboratory\n",
      "323 \t http://vivoweb.org/ontology/core#Location\n",
      "323 \t http://vivoweb.org/ontology/core#GeographicRegion\n",
      "323 \t http://vivoweb.org/ontology/core#GeographicLocation\n",
      "323 \t http://purl.obolibrary.org/obo/BFO_0000141 (immaterial entity)\n",
      "323 \t http://purl.obolibrary.org/obo/BFO_0000006 (spatial region)\n",
      "317 \t http://vivoweb.org/ontology/core#Institute\n",
      "308 \t http://aims.fao.org/aos/geopolitical.owl#area\n",
      "296 \t http://vivoweb.org/ontology/core#GeopoliticalEntity\n",
      "239 \t http://aims.fao.org/aos/geopolitical.owl#territory\n",
      "210 \t http://vivoweb.org/ontology/core#Country\n",
      "200 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#Tool\n",
      "150 \t http://www.w3.org/2004/02/skos/core#Concept\n",
      "150 \t http://org.semanticweb.owlapi/error#Error1\n",
      "69 \t http://aims.fao.org/aos/geopolitical.owl#group\n"
     ]
    }
   ],
   "source": [
    "types = dict()\n",
    "for s,p,o in g.triples( (None, RDF.type, None) ) :\n",
    "    os = str(o)\n",
    "    if os in types.keys():\n",
    "        types[os] += 1\n",
    "    else:\n",
    "        types[os] = 1\n",
    "for v,k in sorted( ((v,k) for k,v in types.items()), reverse=True):\n",
    "    print(v,'\\t',term_lookup(k))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Predicate List\n",
    "\n",
    "Count the occurances of predictes in the triples.  Loop over the triples, counting occurances, \n",
    "then sort the resulting table of predicates and their counts by the counts.  Look up each term\n",
    "to convert the term to something more readable if needed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "542717 \t http://www.w3.org/1999/02/22-rdf-syntax-ns#type\n",
      "131157 \t http://vitro.mannlib.cornell.edu/ns/vitro/0.7#mostSpecificType\n",
      "83728 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#subjectSpecies\n",
      "82658 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#sampleId\n",
      "68100 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#developedFrom\n",
      "68100 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#dataFor\n",
      "40150 \t http://vivoweb.org/ontology/core#relates\n",
      "40150 \t http://vivoweb.org/ontology/core#relatedBy\n",
      "22698 \t http://www.w3.org/2000/01/rdf-schema#label\n",
      "18364 \t http://purl.org/ontology/bibo/doi\n",
      "17790 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#citation\n",
      "17790 \t http://vivoweb.org/ontology/core#dateTimeValue\n",
      "17790 \t http://vivoweb.org/ontology/core#dateTimePrecision\n",
      "17790 \t http://vivoweb.org/ontology/core#dateTime\n",
      "17790 \t http://purl.org/ontology/bibo/pmid\n",
      "2503 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#directs\n",
      "2503 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#directedBy\n",
      "2087 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#associationFor\n",
      "2087 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#associatedWith\n",
      "2063 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#summary\n",
      "1960 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#manages\n",
      "1960 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#managedBy\n",
      "1886 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#workbenchLink\n",
      "1372 \t http://www.w3.org/2006/vcard/ns#hasName\n",
      "1372 \t http://www.w3.org/2006/vcard/ns#givenName\n",
      "1372 \t http://www.w3.org/2006/vcard/ns#familyName\n",
      "1372 \t http://purl.obolibrary.org/obo/ARG_2000029 (vcard of)\n",
      "1372 \t http://purl.obolibrary.org/obo/ARG_2000028 (has vcard)\n",
      "1129 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#parentOf\n",
      "1129 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#hasParent\n",
      "1078 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#runnerOf\n",
      "1078 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#runBy\n",
      "1070 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#studyId\n",
      "1070 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#inCollection\n",
      "1070 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#collectionFor\n",
      "1067 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#submitted\n",
      "910 \t http://vitro.mannlib.cornell.edu/ns/vitro/public#mimeType\n",
      "910 \t http://vitro.mannlib.cornell.edu/ns/vitro/public#filename\n",
      "910 \t http://vitro.mannlib.cornell.edu/ns/vitro/public#downloadLocation\n",
      "910 \t http://vitro.mannlib.cornell.edu/ns/vitro/public#directDownloadUrl\n",
      "895 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#developerOf\n",
      "895 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#developedBy\n",
      "819 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#isPIFor\n",
      "819 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#hasPI\n",
      "816 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#projectId\n",
      "613 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#studyType\n",
      "605 \t http://www.w3.org/2006/vcard/ns#hasEmail\n",
      "605 \t http://www.w3.org/2006/vcard/ns#email\n",
      "552 \t http://www.w3.org/2006/vcard/ns#telephone\n",
      "552 \t http://www.w3.org/2006/vcard/ns#hasTelephone\n",
      "455 \t http://vitro.mannlib.cornell.edu/ns/vitro/public#thumbnailImage\n",
      "455 \t http://vitro.mannlib.cornell.edu/ns/vitro/public#mainImage\n",
      "424 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#projectType\n",
      "209 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#tag\n",
      "200 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#homepage\n",
      "192 \t http://aims.fao.org/aos/geopolitical.owl#isInGroup\n",
      "17 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#licenseUrl\n",
      "17 \t http://www.metabolomics.info/ontologies/2019/metabolomics-consortium#licenseType\n",
      "2 \t http://aims.fao.org/aos/geopolitical.owl#isPredecessorOf\n",
      "1 \t http://aims.fao.org/aos/geopolitical.owl#isSuccessorOf\n"
     ]
    }
   ],
   "source": [
    "predicates = dict()\n",
    "for s,p,o in g:\n",
    "    ps = str(p)\n",
    "    if ps in predicates.keys():\n",
    "        predicates[ps] += 1\n",
    "    else:\n",
    "        predicates[ps] = 1\n",
    "for v,k in sorted( ((v,k) for k,v in predicates.items()), reverse=True):\n",
    "    print(v,'\\t',term_lookup(k))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## List the people in the graph\n",
    "\n",
    "Some of the triples in the graph look like:\n",
    "    \n",
    "     <uri>  rdf:type foaf:Person\n",
    "                \n",
    "These triples are assertions that the uri is the uri of a person.  Find those triples and print them.\n",
    "\n",
    "To find them, loop over all the triples and select those matching the pattern shown.  **None** matches any."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://vivo.metabolomics.info/individual/p404 is a person named Johann Gudjonsson\n",
      "https://vivo.metabolomics.info/individual/p188 is a person named Warren Kruger\n",
      "https://vivo.metabolomics.info/individual/p1351 is a person named Xue Shi\n",
      "https://vivo.metabolomics.info/individual/p325 is a person named Hiroshi Tsugawa\n",
      "https://vivo.metabolomics.info/individual/p1016 is a person named Murray R Badger\n",
      "https://vivo.metabolomics.info/individual/p1181 is a person named David P Enot\n",
      "https://vivo.metabolomics.info/individual/p786 is a person named Reda A Ammar\n",
      "https://vivo.metabolomics.info/individual/p902 is a person named Frank Madeo\n",
      "https://vivo.metabolomics.info/individual/p211 is a person named LaiFang Zhou\n",
      "https://vivo.metabolomics.info/individual/p1273 is a person named Mami Okamoto\n",
      "https://vivo.metabolomics.info/individual/p1257 is a person named Mohammad R Nezami Ranjbar\n"
     ]
    }
   ],
   "source": [
    "n = 0\n",
    "for s,p,o in g.triples( (None, RDF.type, FOAF.Person) ):\n",
    "    print (\"%s is a person named %s\"% (s,g.value(s, RDFS.label)))\n",
    "    n += 1\n",
    "    if n > 10:\n",
    "        break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## List names with special characters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A Ben-Hur\n",
      "Alejandro Soto-Gutierrez\n",
      "Alejandro Villar-Briones\n",
      "Alexandre Perera-Lluna\n",
      "Alison Berent-Spillson\n",
      "Allison O'Kell\n",
      "Alvaro Cuadros-Inostroza\n",
      "Amro, Monte Ilaiwy, WIllis\n",
      "Amro, Monte Ilaiwy, Willis\n",
      "Ana Belén Lozano\n",
      "Antonio Julià\n",
      "Archana Sharma-Oates\n",
      "Aurélie Roux\n",
      "Brian O'Neill\n",
      "Brigitte Wägele\n",
      "Bruce McClenathan, M.D.\n",
      "Carole Sztalryd-Woodle\n",
      "Carolina Gonzalez-Riano\n",
      "Chang Su-youne\n",
      "Charmion Cruickshank-Quinn\n",
      "Ching-Tai Chen\n",
      "Chiun-Gung Juo\n",
      "Christian Jäger\n",
      "Christoph Böttcher\n",
      "Coriness Piñeyro-Ruiz\n",
      "Cristina Andres-Lacueva\n",
      "Cristina Andrés-Lacueva\n",
      "Cristina López-Hidalgo\n",
      "Dennis Mook-Kanamori\n",
      "Domingo Barber Hernández\n",
      "Eric B. Haura\n",
      "Etienne A Thévenot\n",
      "Florence Nicolè\n",
      "Fong-Fu Hsu\n",
      "Francesc Fernández-Albert\n",
      "Gabi Kastenmüller\n",
      "Georg Langenkämper\n",
      "Gildas Le Corguillé\n",
      "Gonçalo Amarante Guimarães Pereira\n",
      "Grace O'Maille\n",
      "Grégory Genta-Jouve\n",
      "Hans-Ulrich Häring\n",
      "Harald C Köfeler\n",
      "Henk Henk W.M. Hilhorst\n",
      "Hugo Peña-Cortés\n",
      "Hui-Yin Chang\n",
      "Hyung-Ok Lee\n",
      "Ivana Blaženović\n",
      "Javier Lopez-Lbanez\n",
      "Javier López-Ibáñez\n",
      "Jean-Baptiste Cazier\n",
      "Jean-Charles Portais\n",
      "Jean-François Martin\n",
      "Ji-Cheng Li\n",
      "Ji-Won Park\n",
      "Johann A Gagnon-Bartsch\n",
      "Jon D. Stewart\n",
      "José Eduardo Santos\n",
      "José Luis Iborra\n",
      "Julian Avila-Pacheco\n",
      "Jörg Kobarg\n",
      "Jürgen Hartler\n",
      "Jürgen Zanghellini\n",
      "K Magnus Åberg\n",
      "Kai Dührkop\n",
      "Ke-Shiuan Lynn\n",
      "Kilian Wörmann\n",
      "Lei Bruschweiler-Li\n",
      "Lilla D'Costa\n",
      "Lourdes Ibañez\n",
      "Lourdes Ibáñez\n",
      "Manuel Cánovas\n",
      "Manuel Llinás\n",
      "Mar Garcia-Aloy\n",
      "Marc-Emmanuel Dumas\n",
      "Marie Tremblay-Franco\n",
      "Marko Sysi-Aho\n",
      "Marta Díaz\n",
      "Marta Ramon-Krauel\n",
      "Martin Trötzmüller\n",
      "Martine Hossaert-McKey\n",
      "Maureen Keller-Wood\n",
      "Meden Isaac-Lam\n",
      "Mi-Jeong Lee\n",
      "Miguel A Rodríguez\n",
      "Miguel Rojas-Chertó\n",
      "Miroslava Cuperlović-Culf\n",
      "Mona ElBadawi-Sidhu\n",
      "Monte, Amro Willis, Ilaiwy\n",
      "Márcia Ortiz Mayo Marques\n",
      "Mélanie Pétéra\n",
      "Mónica Chagoyen\n",
      "Natasha Patel-Murray\n",
      "Núria Mach\n",
      "Pearlie K. Burnette\n",
      "Philippe Schmitt-Kopplin\n",
      "Rachel Culp-Hill\n",
      "Rafael Brüschweiler\n",
      "Ramesh S. Ayyala\n",
      "Randi Gant-Branum\n",
      "Randi Gant-branum\n",
      "Raül Tortosa\n",
      "Rhonda Cooper-DeHoff\n",
      "Ricardo Z N Vêncio\n",
      "Richard Münch\n",
      "Rita Loch-Caruso\n",
      "S. Michael Owens\n",
      "Sang-Joon Ahn\n",
      "Sebastian Böcker\n",
      "Serge Léger\n",
      "Shin-Ichi Utsunomiya\n",
      "Silas Granato Villas-Bôas\n",
      "Simon Tavaré\n",
      "Simone Guidetti-Gonzalez\n",
      "Stefan H.E. Kaufmann\n",
      "Steven O'Hara\n",
      "Suvi Silén\n",
      "Søren B Engelsen\n",
      "Terry E. Weymouth\n",
      "Ting-Yi Sung\n",
      "Tomás Pluskal\n",
      "Tracie Hennen-Bierwagen\n",
      "Ulrich L Günther\n",
      "Valerie de Crecy-Lagard\n",
      "Wei-Shou Hu\n",
      "Wen-Lian Hsu\n",
      "Werner Römisch-Margl\n",
      "William E. and Shipra Smoyer and Agrawal\n",
      "Xavier Domingo-Almenara\n",
      "Yang Mao-Draayer\n",
      "Yasmine Kanaan, PhD.\n",
      "Ying-Jr Chen\n",
      "Yong-Mei Cha\n",
      "Zsuzsanna Lipták\n"
     ]
    }
   ],
   "source": [
    "qres = g.query(\n",
    "    \"\"\"\n",
    "    SELECT ?name\n",
    "    WHERE {\n",
    "      ?p a foaf:Person .\n",
    "      ?p rdfs:label ?name .\n",
    "      FILTER(regex(?name, \"[^A-Za-z ]+?\"))\n",
    "    }\n",
    "    ORDER BY ?name\n",
    "    \"\"\", initNs = { \"vivo\": VIVO, \"foaf\": FOAF, \"rdfs\": RDFS, \"bibo\": BIBO, \"m3c\": M3C})\n",
    "\n",
    "for row in qres:\n",
    "    print(\"%s\" % row)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tabulate Species of Studies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "454 Homo sapiens\n",
      "314 Mus musculus\n",
      "49 Rattus norvegicus\n",
      "17 Mycobacterium tuberculosis H37Rv\n",
      "11 Ovis aries\n",
      "7 Macaca mulatta\n",
      "7 Escherichia coli\n",
      "7 Caenorhabditis elegans\n",
      "7 Plasmodium falciparum\n",
      "7 Arabidopsis thaliana\n",
      "7 None\n",
      "6 Drosophila melanogaster\n",
      "6 Saccharomyces cerevisiae\n",
      "6 Bos taurus\n",
      "5 Rhesus Macaque\n",
      "5 Canis lupus familiaris\n",
      "5 Synechocystis sp. PCC6803\n",
      "5 Megalobrama amblycephala\n",
      "4 Metacarcinus magister\n",
      "4 Plasmodium falciparum;Homo sapiens\n",
      "4 Rattus\n",
      "4 Staphylococcus aureus\n",
      "4 Zea mays\n",
      "3 Trypanosoma brucei brucei\n",
      "3 Chlamydomonas reinhardtii\n",
      "3 Sus scrofa\n",
      "3 Equus caballus\n",
      "3 Homo sapiens | Mus musculus\n",
      "3 Danio rerio\n",
      "3 Salmonella enterica/Escherichia coli\n",
      "3 Felis catus\n",
      "2 Mus musculus/Homo sapiens\n",
      "2 Enterococcus faecalis\n",
      "2 Homo sapiens/Rattus norvegicus\n",
      "2 Cricetinae\n",
      "2 Plasma\n",
      "2 Xenopus laevis\n",
      "2 Auxenochlorella protothecoides\n",
      "2 Bacteria\n",
      "2 Mus\n",
      "2 Bacteroides tethaiotamicron\n",
      "2 Phoenix dactylifera\n",
      "2 Gallus gallus\n",
      "2 Streptococcus mutans\n",
      "2 Sinorhizobium meliloti\n",
      "2 N/A\n",
      "2 Synechococcus elongatus PCC7942 cells\n",
      "2 Oryza sativa\n",
      "2 Centroptilum triangulifer/Athorix diptoa/Megaloptera\n",
      "2 Macaca fascicularis\n",
      "2 Apis mellifera\n",
      "2 Mirounga angustirostris\n",
      "2 Oreochromis mossambicus\n",
      "1 Bacillus subtilis\n",
      "1 Marine Plankton\n",
      "1 Various\n",
      "1 Multiple Plant\n",
      "1 Brugia malayi\n",
      "1 Ceratodon purpurues\n",
      "1 Mixed\n",
      "1 Acropora cervicornis\n",
      "1 Oryctolagus cuniculus\n",
      "1 Microbacterium sediminis\n",
      "1 C. transvaalensis\n",
      "1 Salmonella typhimurium\n",
      "1 Vitis vinifera\n",
      "1 Aedes albopictus\n",
      "1 Triticum durum\n",
      "1 Lactobacillus\n",
      "1 Chemical Standards\n",
      "1 Standard\n",
      "1 mycoplasma genomic variants\n",
      "1 Mesocricetus Auratus\n",
      "1 Tursiops truncatus\n",
      "1 Synechococcus\n",
      "1 Mortierella alpina\n",
      "1 Candida albicans\n",
      "1 Tamiasciurus hudsonicus\n",
      "1 Chinese Hamster\n",
      "1 Thalassiosira pseudonana, karenia brevis\n",
      "1 Panicum maximum(Megathyrsis maximus (Jacq.))\n",
      "1 Nesterenkonia flava\n",
      "1 soybean\n",
      "1 Plasmodium Falciparum\n",
      "1 Solanum lycopersicum\n",
      "1 Nicotiana tabacum\n",
      "1 Bacteroides tethaiotamicron, Bateroides ovatus\n",
      "1 Brassica nigra\n",
      "1 Sarcophaga crassipalpis\n",
      "1 Chlorella minutissima 2341\n",
      "1 Ginkgo Biloba\n",
      "1 Treponema pectinovorum\n",
      "1 Neurospora crassa\n",
      "1 Aphaenogaster picea\n",
      "1 Algae\n",
      "1 Homo sapiens/Callithrix jacchus\n",
      "1 Sus domesticus\n",
      "1 Papio hamadryas\n",
      "1 Canis familiaris\n",
      "1 Porphyromonas gingivalis ATCC 33277\n",
      "1 Bombus impatiens\n",
      "1 Acidithiobacillus ferrooxidans\n",
      "1 Toxoplasma gondii\n",
      "1 Propionibacterium acidipropionici\n",
      "1 Triticum aestivum L.\n",
      "1 Synthetic\n",
      "1 Synechococcus sp CC9605\n",
      "1 Quercus ilex\n",
      "1 Synechococcus sp. PCC 7002\n",
      "1 Callinectus sapidus\n"
     ]
    }
   ],
   "source": [
    "qres = g.query(\"\"\"\n",
    "    SELECT (COUNT (DISTINCT ?s) AS ?nstudies) ?o\n",
    "    WHERE {\n",
    "      ?s a m3c:Study .\n",
    "      ?s m3c:subjectSpecies ?o .\n",
    "    }\n",
    "    GROUP BY ?o\n",
    "    ORDER BY DESC(?nstudies)\n",
    "    \"\"\", initNs = { \"vivo\": VIVO, \"foaf\": FOAF, \"rdfs\": RDFS, \"bibo\": BIBO, \"m3c\": M3C})\n",
    "\n",
    "for row in qres:\n",
    "    print(\"%s %s\" % row)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## List the first 25 publications by DOI and title"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10.20960/nh.591 \n",
      "10.1093/gerona/gln062 \"Dividends\" from research on aging--can biogerontologists, at long last, find something useful to do?\n",
      "http://dx.doi.org/10.21228/M8997F \"Evaluating lipid mediator structural complexity using ion mobility spectrometry combined with mass spectrometry\"\n",
      "10.1016/j.semarthrit.2013.12.007 \"Generalized osteoarthritis\": a systematic review.\n",
      "10.1093/jbcr/irz210 \"Geospatial Mapping as a Guide for Resource Allocation Among Burn Centers in India\".\n",
      "10.1016/j.chembiol.2011.06.001 \"Going KiNativ\": probing the Native Kinome.\n",
      "10.1021/pr400710q \"Out-gel\" tryptic digestion procedure for chemical cross-linking studies with mass spectrometric detection.\n",
      "10.1126/science.aal4573 \"Pheno\"menal value for human health.\n",
      "10.1016/j.hrthm.2014.10.039 \"Power-on resets\" in cardiac implantable electronic devices during magnetic resonance imaging.\n",
      "10.1371/journal.pone.0010936 \"Topological significance\" analysis of gene expression and proteomic profiles from prostate cancer cells reveals key mechanisms of androgen response.\n",
      "10.1016/j.jcmgh.2019.09.009 \"Western Diet\"-Induced Adipose Inflammation Requires a Complex Gut Microbiota.\n",
      "10.1111/j.1474-9728.2004.00081.x 'Accelerated aging': a primrose path to insight?\n",
      "10.1042/BST20160073 'Nothing of chemistry disappears in biology': the Top 30 damage-prone endogenous metabolites.\n",
      "10.5213/inj.2012.16.4.159 'Omics' approaches to understanding interstitial cystitis/painful bladder syndrome/bladder pain syndrome.\n",
      "10.1042/BJ20091328 'Unknown' proteins and 'orphan' enzymes: the missing half of the engineering parts list--and how to find it.\n",
      "10.1016/j.pbb.2004.10.006 (+)-Methamphetamine-induced spontaneous behavior in rats depends on route of (+)METH administration.\n",
      "10.1021/pr900386f (1)H NMR based metabonomics of human amniotic fluid for the metabolic characterization of fetus malformations.\n",
      "10.1016/j.cbi.2014.09.006 (1)H NMR-based metabolomics study on a goldfish model of Parkinson's disease induced by 1-methyl-4-phenyl-1,2,3,6-tetrahydropyridine (MPTP).\n",
      "10.1002/mrc.1764 (1)H relaxation times of metabolites in biological samples obtained with nondestructive ex-vivo slow-MAS NMR.\n",
      "10.1007/s12104-009-9172-6 (1)H, (13)C, and (15)N NMR assignments for the Bacillus subtilis yndB START domain.\n",
      "10.1038/bjc.2015.414 (1)H-NMR urinary metabolomic profiling for diagnosis of gastric cancer.\n",
      "10.1016/j.jprot.2014.06.014 (14)N(15)N DXMSMS Match program for the automated analysis of LC/ESI-MS/MS crosslinking data from experiments using (15)N metabolically labeled proteins.\n",
      "10.1093/jpids/piv014 (1→3)-β-d-Glucan in Cerebrospinal Fluid as a Biomarker for Candida and Aspergillus Infections of the Central Nervous System in Pediatric Patients.\n",
      "10.1093/toxsci/kfv015 1,3-dinitrobenzene induces age- and region-specific oxidation to mitochondria-related proteins in brain.\n",
      "10.1016/j.scitotenv.2019.06.443 1,4-Dioxane as an emerging water contaminant: State of the science and evaluation of research needs.\n"
     ]
    }
   ],
   "source": [
    "qres = g.query(\n",
    "    \"\"\"\n",
    "    SELECT ?doi ?label\n",
    "    WHERE {\n",
    "      ?p bibo:doi ?doi .\n",
    "      ?p rdfs:label ?label .\n",
    "    }\n",
    "    ORDER BY ?label\n",
    "    LIMIT 25\n",
    "    \"\"\", initNs = { \"vivo\": VIVO, \"foaf\": FOAF, \"rdfs\": RDFS, \"bibo\": BIBO})\n",
    "\n",
    "for row in qres:\n",
    "    print(\"%s %s\" % row)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write Graph to a file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f = open(\"out.ttl\", \"w\")\n",
    "print(g.serialize(format=\"ttl\").decode('utf-8'), file=f)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
